crates/kumod/src/egress_source.rs

```diff
@@ -973,17 +1030,152 @@ impl<'a> ProxyProto<'a> {
                     _ => anyhow::bail!("failed to connect {source:?} -> {destination} via {self:?}: {connect_status:?}"),
                 }
 
-                Ok(MaybeProxiedSourceAddress {
-                    address: socks_response_addr(&connect_status)?.into(),
-                    server: Some(server),
-                    protocol: Some("socks5".into()),
-                })
+                let protocol = if tls_enabled { "socks5+tls" } else { "socks5" };
+
+                Ok((
+                    stream,
+                    MaybeProxiedSourceAddress {
+                        address: socks_response_addr(&connect_status)?.into(),
+                        server: Some(server),
+                        protocol: Some(protocol.into()),
+                    },
+                ))
             }
-            Self::None => Ok(MaybeProxiedSourceAddress {
-                address: stream.local_addr()?.into(),
-                server: None,
-                protocol: None,
-            }),
+            Self::None => {
+                let addr = stream.local_addr()?;
+                Ok((
+                    Box::new(stream),
+                    MaybeProxiedSourceAddress {
+                        address: addr.into(),
+                        server: None,
+                        protocol: None,
+                    },
+                ))
+            }
+        }
+    }
+}
+
+/// Build a TLS connector for SOCKS5 proxy connections.
+async fn build_socks5_tls_connector(
```

Comment:
I'd like to avoid copying and pasting the other existing and honestly rather *complex* tls code here, as it will lead to subtle divergence issues in the future.

What I propose is extracting `crates/rfc5321/src/tls.rs` into a separate crate, eg: `crates/kumo-tls-helper`.

I think that `crates/rfc5321/src/traits.rs` should probably also be moved there, as the two are coupled.

Then the logic here for the socks client should be re-phrased to work in terms of the `TlsOptions` struct exposed by that new `kumo-tls-helper` crate.